sdk/sync/types.go

```diff
@@ -1,43 +1,85 @@
 package sync
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"reflect"
-	"strings"
+
+	"github.com/ipfs/testground/sdk/runtime"
 )
 
-// A Subtree represents a subtree of the sync tree for a test run. It is bound
-// to a particular subpath pattern, where payloads of a specific type are posted
-// and manipulated.
-type Subtree struct {
-	// GroupKey is the name of the group the keys in this subtree belong to.
-	GroupKey string
+type State string
+
+// Key gets the Redis key, contextualized to a set of RunParams.
+func (s State) Key(rp *runtime.RunParams) string {
+	p := fmt.Sprintf("run:%s:plan:%s:case:%s:states:%s", rp.TestRun, rp.TestPlan, rp.TestCase, string(s))
+	return p
+}
 
-	// PayloadType is the type of the payload. Must be a pointer type.
-	PayloadType reflect.Type
+type Barrier struct {
+	C chan error
 
-	// KeyFunc returns the key of this entry within the subtree, optionally
-	// deriving it from the supplied payload.
-	KeyFunc func(payload interface{}) string
+	ctx    context.Context
+	state  State
+	key    string
+	target int64
 }
 
-// State represents a state in a distributed test.
-type State string
+type Topic struct {
+	Name string
+	Type reflect.Type
```

Comment:
If we're changing this, consider using:

```go
type Topic struct {
  Name string
  Type interface{}
}

Topic{Name: "foo", Type: new(TypeFoo)}
```

Reply:
Not critical but it removes some reflection from user code.

Reply:
Oh yeah. I guess the Topic can have a sync.Once field to resolve the type and cache it, i.e. to reflect only once. Or I can add a Topic constructor. Good idea!

Reply:
SGTM. Just make sure it's worth it. It looks like we're using reflection every time we emit an event.